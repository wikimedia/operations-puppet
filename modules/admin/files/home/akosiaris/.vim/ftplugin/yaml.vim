" overrides for Debian shipped vim options
setlocal smarttab expandtab shiftwidth=2 softtabstop=2 tabstop=2
