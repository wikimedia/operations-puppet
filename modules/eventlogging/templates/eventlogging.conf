; supervisor configuration for EventLogging
; -----------------------------------------
; This file is managed by Puppet.

; Data flow
; ---------
; - varnishncsa broadcasts UDP log data to vanadium on port 8422.
; - MediaWiki broadcasts UDP log data to vanadium on port 8421.
; - On Vanadium, each incoming UDP stream is re-broadcast over ZMQ (same port
;   number, but TCP/IP). This is to allow for multiple consumers with buffering
;   support.
; - zmq2log consumes each stream and writes it unmodified to disk.
; - log2json consumes the stream from Varnish and transforms it into a parsed
;   and validated JSON stream, which it publishes on port 8522.
; - json2sql consumes the JSON event stream and inserts each event into a MySQL
;   database.

; Logging
; -------
; - Each program's stderr is merged into its stdout (2>&1).
; - stderr log is disabled (per above).
; - stdout log is written to /var/log/supervisor/(program name).log.
; - Log files are rotated at 50MB (default).
; - Ten most recent log archives are kept (default).

; Notifications
; -------------
; - Changes in process status are reported to
;     <eventlogging-alerts@lists.wikimedia.org>.


; UDP -> ZMQ bridge
; -----------------

[program:udp2zmq]
user = eventlogging
numprocs = 2
numprocs_start = 8421
command = udp2zmq %(process_num)d
redirect_stderr = true
process_name = %(program_name)s_%(process_num)d
stdout_logfile = /var/log/supervisor/%(program_name)s_%(process_num)s.log
stderr_logfile = NONE


; File-based logs
; ---------------

[group:zmq2log]
programs = zmq2log-8421,zmq2log-8422,zmq2log-8600
priority = 200

[program:zmq2log-8421]
; Log MediaWiki-generated events (port 8421)
user = eventlogging
command = zmq2log --sid=8421 tcp://127.0.0.1:8421
    /var/log/eventlogging/server-side-events.log
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

[program:zmq2log-8422]
; Log client-side events (port 8422)
user = eventlogging
command = zmq2log --sid=8422 tcp://127.0.0.1:8422
    /var/log/eventlogging/client-side-events.log
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

[program:zmq2log-8600]
; Log all parsed events (port 8600)
user = eventlogging
command = zmq2log --sid=8600 tcp://127.0.0.1:8600
    /var/log/eventlogging/all-events-json.log
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE


; Raw logs -> Parsed / validated JSON
; -----------------------------------

[program:log2json-8521]
user = eventlogging
command = log2json --sid=log2json "%%n\sEventLogging\s%%j" tcp://localhost:8421 tcp://*:8521
priority = 300
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

[program:log2json-8522]
user = eventlogging
command = log2json --sid=log2json "%%q\s%%l\s%%n\s%%t\s%%h" tcp://localhost:8422 tcp://*:8522
priority = 300
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE


; Multiplexing
; ------------

[program:zmux-8600]
user = eventlogging
command = zmux --sid=%(program_name)s tcp://*:8600 tcp://localhost:8521 tcp://localhost:8522
priority = 400
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE


; JSON logs -> MySQL
; ------------------

[program:json2sql-db1047]
user = eventlogging
command = json2sql --sid=db1047 tcp://localhost:8600
    mysql://<%= scope.lookupvar('passwords::mysql::eventlogging::user') %>:<%= scope.lookupvar('passwords::mysql::eventlogging::password') %>@db1047.eqiad.wmnet/log?charset=utf8
priority = 500
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

; JSON logs -> MongoDB
; --------------------

[program:json2mongo-vanadium]
user = eventlogging
command = json2mongo --sid=json2mongo tcp://localhost:8600
    mongodb://vanadium.eqiad.wmnet:27017/?w=1
priority = 500
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

; UDP sequence ID monitoring
; --------------------------

[group:seqmon]
programs = seqmon-8421,seqmon-8422
priority = 999

[program:seqmon-8421]
user = eventlogging
command = seqmon --sid=seqmon --hpos=1 --spos=0 tcp://127.0.0.1:8421
    /var/log/eventlogging/packet-loss-8421.log
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE

[program:seqmon-8422]
user = eventlogging
command = seqmon --sid=seqmon --hpos=1 --spos=2 tcp://127.0.0.1:8422
    /var/log/eventlogging/packet-loss-8422.log
redirect_stderr = true
stdout_logfile = /var/log/supervisor/%(program_name)s.log
stderr_logfile = NONE


; Alerts
; ------

[eventlistener:email-alerts]
command = sv-alerts eventlogging@vanadium.eqiad.wmnet eventlogging-alerts@lists.wikimedia.org
events = PROCESS_STATE_EXITED, PROCESS_STATE_RUNNING
environment = PYTHONUNBUFFERED=1
