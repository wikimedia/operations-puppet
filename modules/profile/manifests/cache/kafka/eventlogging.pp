# SPDX-License-Identifier: Apache-2.0
# === Class profile::cache::kafka::eventlogging
#
# Sets up a varnishkafka logging endpoint for collecting
# analytics events coming from external clients.
#
# More info: https://wikitech.wikimedia.org/wiki/Analytics/EventLogging
#
# === Parameters
#
# [*ensure*]
#   Default: absent.
#   We are removing for https://phabricator.wikimedia.org/T238230.
#   TODO: remove this class and other related code once resources have been ensured absent.
#
# [*kafka_cluster_name*]
#   Name of the Kafka cluster in the kafka_clusters hash to be passed to the
#   kafka_config() function.
#
# [*ssl_enabled*]
#   If true, the Kafka cluster needs to be configured with SSL support.
#   profile::cache::kafka::certificate will be included, and certs used from it.
#   Default: false
#
# [*monitoring_enabled*]
#   True if the varnishkafka instance should be monitored.  Default: false
#
class profile::cache::kafka::eventlogging(
    Wmflib::Ensure $ensure = lookup('profile::cache::kafka::eventlogging::ensure', {'default_value' => 'absent'}),
    String $kafka_cluster_name = lookup('profile::cache::kafka::eventlogging::kafka_cluster_name'),
    Boolean $ssl_enabled = lookup('profile::cache::kafka::eventlogging::ssl_enabled', {'default_value' => false}),
    Boolean $monitoring_enabled = lookup('profile::cache::kafka::eventlogging::monitoring_enabled', {'default_value' => false}),
) {
    $kafka_config = kafka_config($kafka_cluster_name)

    if $ssl_enabled {
        $kafka_brokers = $kafka_config['brokers']['ssl_array']

        include profile::cache::kafka::certificate

        # Include this class to get key and certificate for varnishkafka
        # to produce to Kafka over SSL/TLS.
        $ssl_ca_location = $profile::cache::kafka::certificate::ssl_ca_location
        $ssl_cipher_suites = $profile::cache::kafka::certificate::ssl_cipher_suites
        $ssl_curves_list = $profile::cache::kafka::certificate::ssl_curves_list
        $ssl_sigalgs_list = $profile::cache::kafka::certificate::ssl_sigalgs_list
        $ssl_keystore_location = $profile::cache::kafka::certificate::ssl_keystore_location
        $ssl_keystore_password = $profile::cache::kafka::certificate::ssl_key_password
        $ssl_key_password = $profile::cache::kafka::certificate::ssl_key_password
        $ssl_key_location = $profile::cache::kafka::certificate::ssl_key_location
        $ssl_certificate_location = $profile::cache::kafka::certificate::ssl_certificate_location
    }
    else {
        $kafka_brokers = $kafka_config['brokers']['array']

        $ssl_ca_location = undef
        $ssl_key_password = undef
        $ssl_key_location = undef
        $ssl_certificate_location = undef
        $ssl_cipher_suites = undef
        $ssl_curves_list = undef
        $ssl_sigalgs_list = undef
        $ssl_keystore_location = undef
        $ssl_keystore_password = undef
    }


    # Sometimes we get junk data sent from bunk user agents.
    # Any user agent regex patterns we want to exclude from eventlogging data can be put in this
    # list, and varnishkafka will be configured not to send this data.
    # See: https://phabricator.wikimedia.org/T266130
    $user_agent_exclude_pattern = '^Fuzz Faster U Fool'

    varnishkafka::instance { 'eventlogging':
        ensure                      => $ensure,
        brokers                     => $kafka_brokers,
        # Note that this format uses literal tab characters.
        format                      => '%q	%l	%n	%{%FT%T}t	%{X-Client-IP}o	"%{User-agent}i"',
        format_type                 => 'string',
        compression_codec           => 'snappy',
        topic                       => 'eventlogging-client-side',
        varnish_name                => 'frontend',
        varnish_svc_name            => 'varnish-frontend',
        # Only listen and log requests to /beacon/event(.gif)? that are not from user agents we want to exclude.
        varnish_opts                => { 'q' => "ReqURL ~ \"^/(beacon/)?event(\\.gif)?\\?\" and ReqHeader:user-agent !~ \"${user_agent_exclude_pattern}\"" },
        topic_request_required_acks => '1',
        #TLS/SSL config
        ssl_enabled                 => $ssl_enabled,
        ssl_ca_location             => $ssl_ca_location,
        ssl_key_password            => $ssl_key_password,
        ssl_key_location            => $ssl_key_location,
        ssl_certificate_location    => $ssl_certificate_location,
        ssl_cipher_suites           => $ssl_cipher_suites,
        ssl_curves_list             => $ssl_curves_list,
        ssl_sigalgs_list            => $ssl_sigalgs_list,
        ssl_keystore_location       => $ssl_keystore_location,
        ssl_keystore_password       => $ssl_keystore_password,
    }

    if $monitoring_enabled {
        # Aggregated alarms for delivery errors are defined in icinga::monitor::analytics

        # Generate icinga alert if varnishkafka is not running.
        nrpe::monitor_service { 'varnishkafka-eventlogging':
            ensure        => $ensure,
            description   => 'eventlogging Varnishkafka log producer',
            nrpe_command  => "/usr/lib/nagios/plugins/check_procs -c 1:1 -a '/usr/bin/varnishkafka -S /etc/varnishkafka/eventlogging.conf'",
            contact_group => 'admins,analytics,team-data-platform',
            require       => Varnishkafka::Instance['eventlogging'],
            notes_url     => 'https://wikitech.wikimedia.org/wiki/Analytics/Systems/Varnishkafka',
        }
    }

    # Make sure varnishes are configured and started for the first time
    # before the instances as well, or they fail to start initially...
    Service <| tag == 'varnish_instance' |> -> Varnishkafka::Instance['eventlogging']
}
