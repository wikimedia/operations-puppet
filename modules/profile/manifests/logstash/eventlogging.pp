class profile::logstash::eventlogging {
    require ::profile::logstash::collector

    $topic = 'eventlogging_EventError'
    $kafka_config = kafka_config('analytics')

    logstash::input::kafka { $topic:
        tags              => [$topic, 'kafka'],
        type              => 'eventlogging',
        bootstrap_servers => $kafka_config['brokers']['string'],
    }

    logstash::conf { 'filter_eventlogging':
        source   => 'puppet:///modules/profile/logstash/filter-eventlogging.conf',
        priority => 50,
    }

}
