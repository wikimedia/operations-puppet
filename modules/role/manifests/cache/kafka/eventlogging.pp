class role::cache::kafka::eventlogging(
    $varnish_name = $::hostname,
    $varnish_svc_name = 'varnish',
) inherits role::cache::kafka
{
    # Set varnish.arg.q or varnish.arg.m according to Varnish version
    $varnish_opts = { 'q' => 'ReqURL ~ "^/(beacon/)?event(\.gif)?\?"' }
    $conf_template = 'varnishkafka/varnishkafka_v4.conf.erb'

    varnishkafka::instance { 'eventlogging':
        # FIXME - top-scope var without namespace, will break in puppet 2.8
        # lint:ignore:variable_scope
        brokers                     => $kafka_brokers,
        # lint:endignore
        # Note that this format uses literal tab characters.
        # The '-' in this string used to be %{X-Client-IP@ip}o.
        # EventLogging clientIp logging has been removed as part of T128407.
        format                      => '%q	%l	%n	%{%FT%T}t	-	"%{User-agent}i"',
        format_type                 => 'string',
        topic                       => 'eventlogging-client-side',
        varnish_name                => $varnish_name,
        varnish_svc_name            => $varnish_svc_name,
        varnish_opts                => $varnish_opts,
        topic_request_required_acks => '1',
        conf_template               => $conf_template,
    }
}
