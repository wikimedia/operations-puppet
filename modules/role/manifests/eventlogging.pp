# role/eventlogging.pp
#
# TODO: Move these into role module as
# role::eventlogging::analytics::* classes
#
# These role classes configure various eventlogging services for
# processing analytics EventLogging data.
# The setup is described in detail on
# <https://wikitech.wikimedia.org/wiki/EventLogging>. End-user
# documentation is available in the form of a guide, located at
# <https://www.mediawiki.org/wiki/Extension:EventLogging/Guide>.
#
# There exist two APIs for generating events: efLogServerSideEvent() in
# PHP and mw.eventLog.logEvent() in JavaScript. Events are URL-encoded
# and sent to our servers by means of an HTTP/S request to varnish,
# where a varnishkafka instance forwards to Kafka.  These event streams are
# parsed, validated, and multiplexed into an output streams in Kafka.


# == Class role::eventlogging
# Parent class for eventlogging service role classes.
# This just installs eventlogging and sets up some configuration variables.
#
class role::eventlogging {
    system::role { 'role::eventlogging':
        description => 'EventLogging',
    }

    # EventLogging for analytics processing is deployed
    # as the eventlogging/analytics scap target.
    # eventlogging::deployment::target { 'analytics': }
    # class { 'eventlogging::server':
        # eventlogging_path => '/srv/deployment/eventlogging/analytics'
    # }

    # Get the Kafka configuration
    $kafka_config         = kafka_config('analytics')
    $kafka_brokers_string = $kafka_config['brokers']['string']

    # Where possible, if this is set, it will be included in client configuration
    # to avoid having to do API version for Kafka < 0.10 (where there is not a version API).
    $kafka_api_version = $kafka_config['api_version']

    # Using kafka-confluent as a consumer is not currently supported by this puppet module,
    # but is implemented in eventlogging.  Hardcode the scheme for consumers for now.
    $kafka_consumer_scheme = 'kafka://'

    # Commonly used Kafka input URIs.
    $kafka_mixed_uri = "${kafka_consumer_scheme}/${kafka_brokers_string}?topic=eventlogging-valid-mixed"
    $kafka_client_side_raw_uri = "${kafka_consumer_scheme}/${kafka_brokers_string}?topic=eventlogging-client-side"

    # # This check was written for eventlog1001, so only include it there.,
    # if $::hostname == 'eventlog1001' {
    #
    #     # Alert when / gets low. (eventlog1001 has a 9.1G /)
    #     nrpe::monitor_service { 'eventlogging_root_disk_space':
    #         description   => 'Eventlogging / disk space',
    #         nrpe_command  => '/usr/lib/nagios/plugins/check_disk -w 1024M -c 512M -p /',
    #         contact_group => 'analytics',
    #     }
    #
    #     # Alert when /srv gets low. (eventlog1001 has a 456G /srv)
    #     # Currently, /srv/log/eventlogging grows at about 500kB / s.
    #     # Which is almost 2G / hour.  100G gives us about 2 days to respond,
    #     # 50G gives us about 1 day.  Logrotate should keep enough disk space free.
    #     nrpe::monitor_service { 'eventlogging_srv_disk_space':
    #         description   => 'Eventlogging /srv disk space',
    #         nrpe_command  => '/usr/lib/nagios/plugins/check_disk -w 100000M -c 50000M -p /srv',
    #         contact_group => 'analytics',
    #     }
    # }
    #
    # # make sure any defined eventlogging services are running
    # class { '::eventlogging::monitoring::jobs': }
}


#
# ==== Data flow classes ====
#


# == Class role::eventlogging::forwarder
# This class was responsible for forwarding incoming raw events from UDP
# into the eventlogging system - this behavior has been deprecated and
# server side events are directly sent from Mediawiki via HTTP POST to the
# event beacon endpoint and will be processed as client side events.
#
# This forwarder class exists only for backwards compatibility for services
# consuming from the legacy ZMQ stream now.
class role::eventlogging::forwarder inherits role::eventlogging {
    $eventlogging_host    = hiera('eventlogging_host', $::ipaddress)

    # This forwards the kafka eventlogging-valid-mixed topic to
    # ZMQ port 8600 for backwards compatibility.
    eventlogging::service::forwarder { 'legacy-zmq':
        input   => "${kafka_mixed_uri}&enable_auto_commit=False&identity=eventlogging_legacy_zmq",
        outputs => ["tcp://${eventlogging_host}:8600"],
    }

    ferm::service { 'eventlogging-zmq-legacy-stream':
        proto   => 'tcp',
        notrack => true,
        port    => '8600',
        srange  => '@resolve((hafnium.eqiad.wmnet graphite1001.eqiad.wmnet))',
    }
}


# == Class role::eventlogging::processor
# Reads raw events, parses and validates them, and then sends
# them along for further consumption.
#
class role::eventlogging::processor inherits role::eventlogging {
    $kafka_consumer_group = hiera(
        'eventlogging_processor_kafka_consumer_group',
        'eventlogging_processor_client_side_00'
    )

    # Run N parallel client side processors.
    # These will auto balance amongst themselves.
    $client_side_processors = hiera(
        'eventlogging_client_side_processors',
        ['client-side-00', 'client-side-01']
    )

    # Choose the eventlogging URI scheme to use for consumers and producer (inputs vs outputs).
    # This allows us to try out different Kafka handlers and different kafka clients
    # that eventlogging supports.  The default is kafka://.  Also available is kafka-confluent://
    # eventlogging::processor is the only configured analytics eventlogging kafka producer, so we
    # only need to define this here.
    $kafka_producer_scheme = hiera('eventlogging_kafka_producer_scheme', 'kafka://')

    # Read in raw events from Kafka, process them, and send them to
    # the schema corresponding to their topic in Kafka.
    $kafka_schema_output_uri  = "${kafka_producer_scheme}/${kafka_brokers_string}?topic=eventlogging_{schema}"

    # The downstream eventlogging MySQL consumer expects schemas to be
    # all mixed up in a single stream.  We send processed events to a
    # 'mixed' kafka topic in order to keep supporting it for now.
    # We blacklist certain high volume schemas from going into this topic.
    $mixed_schema_blacklist = hiera('eventlogging_valid_mixed_schema_blacklist', undef)
    $kafka_mixed_output_uri = $mixed_schema_blacklist ? {
        undef   => "${kafka_producer_scheme}/${kafka_brokers_string}?topic=eventlogging-valid-mixed",
        default => "${kafka_producer_scheme}/${kafka_brokers_string}?topic=eventlogging-valid-mixed&blacklist=${mixed_schema_blacklist}"
    }

    # Append this to query params of kafka-python writer if set.
    # kafka-confluent defaults to setting this to 0.9 anyway.
    $kafka_api_version_param = $kafka_api_version ? {
        undef   => '',
        default => "&api_version=${kafka_api_version}"
    }

    # Increase number and backoff time of retries for async
    # analytics uses.  If metadata changes, we should give
    # more time to retry. NOTE: testing this in production
    # STILL yielded some dropped messages.  Need to figure
    # out why and stop it.  This either needs to be higher,
    # or it is a bug in kafka-python.
    # See: https://phabricator.wikimedia.org/T142430
    $kafka_producer_args = $kafka_producer_scheme ? {
        # args for kafka-confluent handler writer
        'kafka-confluent://' => 'message.send.max.retries=6,retry.backoff.ms=200',
        # args for kafka-python handler writer
        'kafka://'           => "retries=6&retry_backoff_ms=200${kafka_api_version_param}"
    }

    eventlogging::service::processor { $client_side_processors:
        format         => '%q %{recvFrom}s %{seqId}d %t %o %{userAgent}i',
        input          => "${kafka_client_side_raw_uri}${kafka_api_version_param}",
        sid            => $kafka_consumer_group,
        outputs        => [
            "${kafka_schema_output_uri}&${kafka_producer_args}",
            "${kafka_mixed_output_uri}&${kafka_producer_args}",
        ],
        output_invalid => true,
    }
}


# == Class role::eventlogging::consumer::mysql
# Consumes the stream of events and writes them to MySQL
#
class role::eventlogging::consumer::mysql inherits role::eventlogging {
    ## MySQL / MariaDB

    # Log strictly valid events to the 'log' database on m4-master.

    class { 'passwords::mysql::eventlogging': }    # T82265
    $mysql_user = $passwords::mysql::eventlogging::user
    $mysql_pass = $passwords::mysql::eventlogging::password
    $mysql_db = $::realm ? {
        production => 'm4-master.eqiad.wmnet/log',
        labs       => '127.0.0.1/log',
    }

    # Run N parallel mysql consumers processors.
    # These will auto balance amongst themselves.
    $mysql_consumers = hiera(
        'eventlogging_mysql_consumers',
        ['mysql-m4-master-00']
    )
    $kafka_consumer_group = 'eventlogging_consumer_mysql_00'

    # Append this to query params if set.
    $kafka_api_version_param = $kafka_api_version ? {
        undef   => '',
        default => "&api_version=${kafka_api_version}"
    }

    # Define statsd host url to send mysql insert metrics.
    # For beta cluster, set in https://wikitech.wikimedia.org/wiki/Hiera:Deployment-prep
    $statsd_host          = hiera('eventlogging_statsd_host',      'statsd.eqiad.wmnet')

    # Kafka consumer group for this consumer is mysql-m4-master
    eventlogging::service::consumer { $mysql_consumers:
        # auto commit offsets to kafka more often for mysql consumer
        input  => "${kafka_mixed_uri}&auto_commit_interval_ms=1000${$kafka_api_version_param}",
        output => "mysql://${mysql_user}:${mysql_pass}@${mysql_db}?charset=utf8&statsd_host=${statsd_host}&replace=True",
        sid    => $kafka_consumer_group,
        # Restrict permissions on this config file since it contains a password.
        owner  => 'root',
        group  => 'eventlogging',
        mode   => '0640',
    }
}


# == Class role::eventlogging::consumer::files
# Consumes streams of events and writes them to log files.
#
class role::eventlogging::consumer::files inherits role::eventlogging {
    # Log all raw log records and decoded events to flat files in
    # $out_dir as a medium of last resort. These files are rotated
    # and rsynced to stat1003 & stat1002 for backup.

    $out_dir = '/srv/log/eventlogging'

    # We ensure the /srv/log (parent of $out_dir) manually here, as
    # there is no proper class to rely on for this, and starting a
    # separate would be an overkill for now.
    if !defined(File['/srv/log']) {
        file { '/srv/log':
            ensure => 'directory',
            mode   => '0755',
            owner  => 'root',
            group  => 'root',
        }
    }

    # Logs are collected in <$log_dir> and rotated daily.
    file { [$out_dir, "${out_dir}/archive"]:
        ensure => 'directory',
        owner  => 'eventlogging',
        group  => 'eventlogging',
        mode   => '0664',
    }

    # TODO put this in a file when role::eventlogging
    # moves to role modules.
    $logrotate_content = "${out_dir}/*.log {
       daily
       olddir ${out_dir}/archive
       notifempty
       maxage 30
       rotate 1000
       dateext
       compress
       delaycompress
       missingok
}"

    logrotate::conf { 'eventlogging-files':
        ensure  => present,
        content => $logrotate_content,
        require => [
            File[$out_dir],
            File["${out_dir}/archive"]
        ],
    }

    $kafka_consumer_group = hiera(
        'eventlogging_files_kafka_consumer_group',
        'eventlogging_consumer_files_00'
    )

    # Append this to query params if set.
    $kafka_api_version_param = $kafka_api_version ? {
        undef   => '',
        default => "&api_version=${kafka_api_version}"
    }

    # Raw client side events:
    eventlogging::service::consumer { 'client-side-events.log':
        input  => "${kafka_client_side_raw_uri}&raw=True${kafka_api_version_param}",
        output => "file://${out_dir}/client-side-events.log",
        sid    => $kafka_consumer_group,
    }
    # Processed and valid all (AKA 'mixed') mixed.
    # Note that this does not include events that were
    # 'blacklisted' during processing.  Events are blacklisted
    # from these logs for volume reasons.
    eventlogging::service::consumer { 'all-events.log':
        input  =>  "${kafka_mixed_uri}${kafka_api_version_param}",
        output => "file://${out_dir}/all-events.log",
        sid    => $kafka_consumer_group,
    }

    $backup_destinations = $::realm ? {
        production => [  'stat1002.eqiad.wmnet', 'stat1003.eqiad.wmnet' ],
        labs       => false,
    }

    if ( $backup_destinations ) {
        class { 'rsync::server': }

        $rsync_clients_ferm = join($backup_destinations, ' ')
        ferm::service { 'eventlogging_rsyncd':
            proto  => 'tcp',
            port   => '873',
            srange => "@resolve((${rsync_clients_ferm}))",
        }

        rsync::server::module { 'eventlogging':
            path        => $out_dir,
            read_only   => 'yes',
            list        => 'yes',
            require     => File[$out_dir],
            hosts_allow => $backup_destinations,
        }
    }
}
