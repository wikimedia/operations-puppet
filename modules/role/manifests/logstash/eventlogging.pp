# vim:sw=4 ts=4 sts=4 et:
# == Class: role::logstash::eventlogging
#
# Configure Logstash to consume validation logs from EventLogging.
#
class role::logstash::eventlogging {
    include ::role::logstash::collector

    $topic = 'eventlogging_EventError'
    $kafka_config = kafka_config('analytics')

    logstash::input::kafka { $topic:
        tags       => [$topic, 'kafka'],
        type       => 'eventlogging',
        zk_connect => $kafka_config['zookeeper']['url'],
    }
    # lint:ignore:puppet_url_without_modules
    logstash::conf { 'filter_eventlogging':
        source   => 'puppet:///modules/role/logstash/filter-eventlogging.conf',
        priority => 50,
    }
    # lint:endignore
}
