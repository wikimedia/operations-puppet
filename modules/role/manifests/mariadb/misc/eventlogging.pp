# Eventlogging needs to be sandboxed by itself. It can consume resources
# unpredictably, especially during backfilling. It also benefits greatly
# from a setup tuned for TokuDB.
class role::mariadb::misc::eventlogging(
    $shard  = 'm4',
    $master = false,
    ) {

    system::role { 'role::mariadb::misc':
        description => 'Eventlogging Database',
    }

    $mysql_role = $master ? {
        true  => 'master',
        false => 'slave',
    }

    include ::standard
    include role::mariadb::monitor::dba
    include passwords::misc::scripts
    include role::mariadb::ferm

    class {'role::mariadb::groups':
        mysql_group => 'misc',
        mysql_shard => $shard,
        mysql_role  => $mysql_role,
    }

    include mariadb::packages_wmf
    include mariadb::service

    $read_only = $master ? {
        true  => 0,
        false => 1,
    }

    class { 'mariadb::config':
        config        => 'role/mariadb/mysqld_config/eventlogging.my.cnf.erb',
        datadir       => '/srv/sqldata',
        tmpdir        => '/srv/tmp',
        read_only     => $read_only,
        ssl           => 'puppet-cert',
        p_s           => 'off',
        binlog_format => 'MIXED',
    }

    class { 'role::mariadb::grants::production':
        shard    => $shard,
        prompt   => "EVENTLOGGING ${shard}",
        password => $passwords::misc::scripts::mysql_root_pass,
    }

    class { 'mariadb::heartbeat':
        shard      => $shard,
        datacenter => $::site,
        enabled    => $master,
    }
}

