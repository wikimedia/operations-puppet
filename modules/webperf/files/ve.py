#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import sys
reload(sys)
sys.setdefaultencoding("utf-8")

import argparse
import socket
import unittest
import yaml

import eventlogging

def handle_edit(meta):
    event = meta['event']
    if event['editor'] == 'visualeditor':
        action = event['action']
        if action == 'saveSuccess':
            metric = 'save'
        elif action == 'ready':
            metric = 'load'
        else:
            return
        timing = int(event['action.%s.timing' % action])
        # Log values between 0.1s an 100s (1.6min) only
        if timing < 100 or timing > 100000:
            return
        return 'VisualEditor.%s:%s|ms' % (metric, timing)


if __name__ == '__main__':
    ap = argparse.ArgumentParser(description='PerfData StatsD module')
    ap.add_argument('endpoint', help='URI of EventLogging endpoint')
    ap.add_argument('--statsd-host', default='localhost',
                    type=socket.gethostbyname)
    ap.add_argument('--statsd-port', default=8125, type=int)
    args = ap.parse_args()

    addr = args.statsd_host, args.statsd_port
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    events = eventlogging.connect(args.endpoint)

    for meta in events.filter(schema='Edit'):
        try:
            stat = handle_edit(meta)
            if stat is not None:
                sock.sendto(stat.encode('utf-8'), addr)
        except (ValueError, KeyError):
            continue


# ##### Tests ######
# To run:
#   python -m unittest ve
#
class TestVePerfData(unittest.TestCase):
    def test_handler(self):
        with open('ve_fixture.yaml') as fixture_file:
            fixture = yaml.safe_load(fixture_file)
            actual = []
            for meta in fixture:
                stat = handle_edit(meta)
                if stat is not None:
                    actual.append(stat)
            with open('ve_expected.txt') as expected_file:
                self.assertItemsEqual(
                    actual,
                    expected_file.read().splitlines()
                )
