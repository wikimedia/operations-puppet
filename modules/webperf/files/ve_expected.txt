VisualEditor.load:1955|ms
VisualEditor.load:33097|ms
VisualEditor.load:3077|ms
VisualEditor.load:2963|ms
VisualEditor.load:1769|ms
VisualEditor.save:14379|ms
VisualEditor.load:1067|ms
VisualEditor.load:9615|ms
VisualEditor.load:2391|ms
VisualEditor.load:16085|ms
VisualEditor.load:2711|ms
